rfcs/063-catalogue_pipeline_lambdas/README.md

```diff
@@ -0,0 +1,139 @@
+
+## Background
+
+The following points cover in more detail the reasoning behind moving from ECS based compute to AWS Lamdba:
+
+- **Reduction in infra complexity by removing ECS auto-scaling logic:**
+
+   Scaling the pipeline up or down is achieved in several ways:
+   
+   - A manual step is performed to terraform the pipeline into "reindexing_state" by setting a bunch of variables to `true` [here](pipeline/terraform/2024-08-15/main.tf). Search for `reindexing_state` in the catalogue-pipeline repo to find all the 
+     ways these variables are used.  
+   - Every service also has a high/low message count alarm on its input queue. By [default](pipeline/terraform/modules/stack/variables.tf) services scale up to 12 instances as long as there are messages on the queue, and will scale down when the queue 
+     is clear. However some services (eg. ingestors) have specially configured `max_capacity` usually to account for data stores' read/write limits.
+
+   This change aims to remove the need for this scaling logic.
+
+- **Reduction in app complexity by removing akka/pekko batching:**
+
+   Several (SQS driven) pipeline services implement “SQSStream” (see https://github.com/wellcomecollection/scala-libs/blob/main/messaging/src/main/scala/weco/messaging/sqs/SQSStream.scala ) that batches SQS messages so that they can be processed in 
+   bulk. This logic lives inside the service and makes it difficult (impossible?) to run locally.  
+
+   This model means the messaging and business/processing logic part of the services are entangled in sometimes confusing ways, and we're losing queue management features of AWS SQS (eg. having to keep track of success and failure to delete or retry 
+   messaging accordingly).  
+ 
+   This change is also an opportunity to reassess and possibly refactor out some error handling that may be too particular, over-optimising for performance at the expense of clarity. 
+
+- **Potential reductions in cost:**
+ 
+   ECS services are configured to scale down to 0 when the input is low, which means we are not paying for continously-running tasks, but this results in startinf and stopping tasks repeatedly as messages trickle down the pipeline.  
+
+- **Potential increase in speed (deployment and processing):**
+
+   ECS services take a few minutes to start tasks as the auto-scaling rules rely on reported CloudWatch metrics, whereas Lambda invocations are [much faster](https://docs.aws.amazon.com/lambda/latest/operatorguide/execution-environments.html#cold-start-latency). Using [provisioned concurrency](https://docs.aws.amazon.com/lambda/latest/dg/provisioned-concurrency.html) may further increase processing speed, but is likely not necessary, at least for most services.
+
+  Service deployment at present relies on the ECS deployment APIs which take some time to determine that a service is "stable", there is no such process for AWS Lambda and deployments are almost instant.
+
```

Comment:
I think another speed aspect is that we currently try to compromise between normal running and reindexing with the way we gather messages.  This, along with the way we pull messages into an internal queue until the application decides it's got enough messages/waited long enough means that we can see artificial delays during normal running, and towards the end of a reindex (e.g. because multiple ECS instances have pulled some of the messages, and now there aren't enough to go around)

Reply:
Yes, i'll rephrase for the document and add this point.